pof/src/types.rs

```diff
@@ -2148,18 +2137,35 @@ impl Model {
         if let Some(&detail_0) = self.header.detail_levels.first() {
             let num_verts = sum_verts_recurse(&self.sub_objects, detail_0);
 
-            let point_mass = self.header.mass / num_verts as f32;
+            fn add_point_mass_moi(moi: &mut Matrix3<f64>, pos: Vec3d) {
+                moi.column_mut(0).x += (pos.y * pos.y + pos.z * pos.z) as f64;
+                moi.column_mut(0).y -= (pos.x * pos.y) as f64;
+                moi.column_mut(0).z -= (pos.x * pos.z) as f64;
+                moi.column_mut(1).x -= (pos.x * pos.y) as f64;
+                moi.column_mut(1).y += (pos.x * pos.x + pos.z * pos.z) as f64;
+                moi.column_mut(1).z -= (pos.y * pos.z) as f64;
+                moi.column_mut(2).x -= (pos.x * pos.z) as f64;
+                moi.column_mut(2).y -= (pos.y * pos.z) as f64;
+                moi.column_mut(2).z += (pos.x * pos.x + pos.y * pos.y) as f64;
+            }
 
-            fn accumulate_moi_recurse(subobjects: &ObjVec<SubObject>, id: ObjectId, moi: &mut Mat3d) {
-                subobjects[id].bsp_data.verts.iter().for_each(|vert| moi.add_point_mass_moi(*vert));
+            fn accumulate_moi_recurse(subobjects: &ObjVec<SubObject>, id: ObjectId, moi: &mut Matrix3<f64>) {
+                subobjects[id].bsp_data.verts.iter().for_each(|vert| add_point_mass_moi(moi, *vert));
                 subobjects[id].children.iter().for_each(|id| accumulate_moi_recurse(subobjects, *id, moi));
             }
 
-            accumulate_moi_recurse(&self.sub_objects, detail_0, &mut self.header.moment_of_inertia);
+            let mut new_moi: Matrix3<f64> = Matrix3::identity();
 
-            let mut glm_mat: Mat3x3 = self.header.moment_of_inertia.into();
-            glm_mat *= point_mass;
-            self.header.moment_of_inertia = glm_mat.try_inverse().unwrap().into();
+            accumulate_moi_recurse(&self.sub_objects, detail_0, &mut new_moi);
+
+            let point_mass = self.header.mass as f64 / num_verts as f64;
+            new_moi *= point_mass;
+            new_moi = new_moi.try_inverse().unwrap();
+            self.header.moment_of_inertia = Mat3d {
+                rvec: Vec3d::new(new_moi.column(0).x as f32, new_moi.column(0).y as f32, new_moi.column(0).z as f32),
+                uvec: Vec3d::new(new_moi.column(1).x as f32, new_moi.column(1).y as f32, new_moi.column(1).z as f32),
+                fvec: Vec3d::new(new_moi.column(2).x as f32, new_moi.column(2).y as f32, new_moi.column(2).z as f32),
+            };
```

Comment:
surelo there's a function for converting to Matrix3<f32>, and then you can use your `into()` to convert to `Mat3x3`